// lineupjs/src/config.js
import { renderers } from './renderer';
import { toolbarActions } from './ui';
export * from './interfaces';
export function defaultOptions() {
    return {
        toolbar: Object.assign({}, toolbarActions),
        renderers: Object.assign({}, renderers),
        labelRotation: 0,
        summaryHeader: true,
        animated: true,
        expandLineOnHover: false,
        sidePanel: true,
        sidePanelCollapsed: false,
        hierarchyIndicator: true,
        defaultSlopeGraphMode: 'item',
        overviewMode: false,
        rowHeight: 18,
        groupHeight: 40,
        groupPadding: 5,
        rowPadding: 2,
        levelOfDetail: function () { return 'high'; },
        customRowUpdate: function () { return undefined; },
        dynamicHeight: function () { return null; },
        flags: {
            disableFrozenColumns: true,
            advancedRankingFeatures: true,
            advancedModelFeatures: true,
            advancedUIFeatures: true
        },
        ignoreUnsupportedBrowser: false
    };
}
