// lineupjs/src/builder/adapter/index.js
import { build, buildCategorical, buildDate, buildHierarchy, buildNumber, buildString, buildActions } from './column';
import { Adapter } from './lineup';
import { buildAllColumnsRanking, buildGeneric, buildImposeRanking, buildNestedRanking, buildRanking, buildReduceRanking, buildScriptRanking, buildSupportRanking, buildWeightedSumRanking } from './ranking';
import { equal, isSame, isTypeInstance } from './utils';
export var builderAdapter = {
    buildString: buildString,
    buildNumber: buildNumber,
    buildHierarchy: buildHierarchy,
    buildDate: buildDate,
    buildCategorical: buildCategorical,
    buildActions: buildActions,
    build: build,
    buildGeneric: buildGeneric,
    buildWeightedSumRanking: buildWeightedSumRanking,
    buildSupportRanking: buildSupportRanking,
    buildScriptRanking: buildScriptRanking,
    buildReduceRanking: buildReduceRanking,
    buildRanking: buildRanking,
    buildNestedRanking: buildNestedRanking,
    buildImposeRanking: buildImposeRanking,
    buildAllColumnsRanking: buildAllColumnsRanking,
    equal: equal,
    isSame: isSame,
    isTypeInstance: isTypeInstance,
    Adapter: Adapter
};
