// lineupjs/src/model/CompositeNumberColumn.js
import { __decorate, __extends } from "tslib";
import { format } from 'd3-format';
import CompositeColumn from './CompositeColumn';
import { isMissingValue } from './missing';
import NumberColumn from './NumberColumn';
import { SortByDefault } from './annotations';
var CompositeNumberColumn = (function (_super) {
    __extends(CompositeNumberColumn, _super);
    function CompositeNumberColumn(id, desc) {
        var _this = _super.call(this, id, desc) || this;
        _this.missingValue = NaN;
        _this.numberFormat = format('.3n');
        if (desc.numberFormat) {
            _this.numberFormat = format(desc.numberFormat);
        }
        if (desc.missingValue !== undefined) {
            _this.missingValue = desc.missingValue;
        }
        return _this;
    }
    CompositeNumberColumn.prototype.getNumberFormat = function () {
        return this.numberFormat;
    };
    CompositeNumberColumn.prototype.dump = function (toDescRef) {
        var r = _super.prototype.dump.call(this, toDescRef);
        r.missingValue = this.missingValue;
        return r;
    };
    CompositeNumberColumn.prototype.restore = function (dump, factory) {
        if (dump.missingValue !== undefined) {
            this.missingValue = dump.missingValue;
        }
        if (dump.numberFormat) {
            this.numberFormat = format(dump.numberFormat);
        }
        _super.prototype.restore.call(this, dump, factory);
    };
    CompositeNumberColumn.prototype.getLabel = function (row) {
        if (!this.isLoaded()) {
            return '';
        }
        var v = this.getValue(row);
        return String(typeof v === 'number' && !isNaN(v) && isFinite(v) ? this.numberFormat(v) : v);
    };
    CompositeNumberColumn.prototype.getValue = function (row) {
        if (!this.isLoaded()) {
            return null;
        }
        var v = this.compute(row);
        if (isMissingValue(v)) {
            return this.missingValue;
        }
        return v;
    };
    CompositeNumberColumn.prototype.compute = function (_row) {
        return NaN;
    };
    CompositeNumberColumn.prototype.getNumber = function (row) {
        var r = this.getValue(row);
        return r == null ? NaN : r;
    };
    CompositeNumberColumn.prototype.getRawNumber = function (row) {
        return this.getNumber(row);
    };
    CompositeNumberColumn.prototype.getExportValue = function (row, format) {
        if (format === 'json') {
            return {
                value: this.getRawNumber(row),
                children: this.children.map(function (d) { return d.getExportValue(row, format); })
            };
        }
        return _super.prototype.getExportValue.call(this, row, format);
    };
    CompositeNumberColumn.prototype.isMissing = function (row) {
        return isMissingValue(this.compute(row));
    };
    CompositeNumberColumn.prototype.compare = function (a, b) {
        return NumberColumn.prototype.compare.call(this, a, b);
    };
    CompositeNumberColumn.prototype.groupCompare = function (a, b) {
        return NumberColumn.prototype.groupCompare.call(this, a, b);
    };
    CompositeNumberColumn.prototype.getRenderer = function () {
        return NumberColumn.prototype.getRenderer.call(this);
    };
    CompositeNumberColumn = __decorate([
        SortByDefault('descending')
    ], CompositeNumberColumn);
    return CompositeNumberColumn;
}(CompositeColumn));
export default CompositeNumberColumn;
