// lineupjs/src/model/NumbersColumn.js
import { __decorate, __extends } from "tslib";
import { LazyBoxPlotData } from '../internal';
import { toolbar, SortByDefault, dialogAddons } from './annotations';
import ArrayColumn from './ArrayColumn';
import Column from './Column';
import { compareBoxPlot, DEFAULT_FORMATTER, EAdvancedSortMethod, getBoxPlotNumber, noNumberFilter, isDummyNumberFilter, restoreNumberFilter } from './INumberColumn';
import { createMappingFunction, restoreMapping, ScaleMappingFunction } from './MappingFunction';
import { isMissingValue } from './missing';
import NumberColumn from './NumberColumn';
import { restoreColorMapping, createColorMappingFunction } from './ColorMappingFunction';
var NumbersColumn = (function (_super) {
    __extends(NumbersColumn, _super);
    function NumbersColumn(id, desc) {
        var _this = _super.call(this, id, desc) || this;
        _this.currentFilter = noNumberFilter();
        _this.mapping = restoreMapping(desc);
        _this.original = _this.mapping.clone();
        _this.colorMapping = restoreColorMapping(desc);
        _this.sort = desc.sort || EAdvancedSortMethod.median;
        if (_this.dataLength) {
            _this.setDefaultWidth(Math.min(Math.max(100, _this.dataLength * 10), 500));
        }
        _this.setDefaultRenderer('heatmap');
        _this.setDefaultGroupRenderer('heatmap');
        _this.setDefaultSummaryRenderer('histogram');
        return _this;
    }
    NumbersColumn_1 = NumbersColumn;
    NumbersColumn.prototype.compare = function (a, b) {
        return compareBoxPlot(this, a, b);
    };
    NumbersColumn.prototype.getRawNumbers = function (row) {
        return this.getRawValue(row);
    };
    NumbersColumn.prototype.getBoxPlotData = function (row) {
        var data = this.getRawValue(row);
        if (data == null) {
            return null;
        }
        return new LazyBoxPlotData(data, this.mapping);
    };
    NumbersColumn.prototype.getRange = function () {
        return this.mapping.getRange(DEFAULT_FORMATTER);
    };
    NumbersColumn.prototype.getRawBoxPlotData = function (row) {
        var data = this.getRawValue(row);
        if (data == null) {
            return null;
        }
        return new LazyBoxPlotData(data);
    };
    NumbersColumn.prototype.getNumbers = function (row) {
        return this.getValue(row);
    };
    NumbersColumn.prototype.getNumber = function (row) {
        return getBoxPlotNumber(this, row, 'normalized');
    };
    NumbersColumn.prototype.getRawNumber = function (row) {
        return getBoxPlotNumber(this, row, 'raw');
    };
    NumbersColumn.prototype.getValue = function (row) {
        var _this = this;
        var values = this.getRawValue(row);
        return values.map(function (d) { return isMissingValue(d) ? NaN : _this.mapping.apply(d); });
    };
    NumbersColumn.prototype.getRawValue = function (row) {
        var r = _super.prototype.getValue.call(this, row);
        return r == null ? [] : r;
    };
    NumbersColumn.prototype.getExportValue = function (row, format) {
        return format === 'json' ? this.getRawValue(row) : _super.prototype.getExportValue.call(this, row, format);
    };
    NumbersColumn.prototype.getLabels = function (row) {
        return this.getValue(row).map(DEFAULT_FORMATTER);
    };
    NumbersColumn.prototype.getSortMethod = function () {
        return this.sort;
    };
    NumbersColumn.prototype.setSortMethod = function (sort) {
        if (this.sort === sort) {
            return;
        }
        this.fire([NumbersColumn_1.EVENT_SORTMETHOD_CHANGED], this.sort, this.sort = sort);
        if (!this.isSortedByMe().asc) {
            this.sortByMe();
        }
    };
    NumbersColumn.prototype.dump = function (toDescRef) {
        var r = _super.prototype.dump.call(this, toDescRef);
        r.sortMethod = this.getSortMethod();
        r.filter = !isDummyNumberFilter(this.currentFilter) ? this.currentFilter : null;
        r.map = this.mapping.dump();
        r.colorMapping = this.colorMapping.dump();
        return r;
    };
    NumbersColumn.prototype.restore = function (dump, factory) {
        _super.prototype.restore.call(this, dump, factory);
        if (dump.sortMethod) {
            this.sort = dump.sortMethod;
        }
        if (dump.filter) {
            this.currentFilter = restoreNumberFilter(dump.filter);
        }
        if (dump.map) {
            this.mapping = createMappingFunction(dump.map);
        }
        else if (dump.domain) {
            this.mapping = new ScaleMappingFunction(dump.domain, 'linear', dump.range || [0, 1]);
        }
        if (dump.colorMapping) {
            this.colorMapping = createColorMappingFunction(dump.colorMapping);
        }
    };
    NumbersColumn.prototype.createEventList = function () {
        return _super.prototype.createEventList.call(this).concat([NumbersColumn_1.EVENT_COLOR_MAPPING_CHANGED, NumbersColumn_1.EVENT_MAPPING_CHANGED, NumbersColumn_1.EVENT_SORTMETHOD_CHANGED, NumbersColumn_1.EVENT_FILTER_CHANGED]);
    };
    NumbersColumn.prototype.on = function (type, listener) {
        return _super.prototype.on.call(this, type, listener);
    };
    NumbersColumn.prototype.getOriginalMapping = function () {
        return this.original.clone();
    };
    NumbersColumn.prototype.getMapping = function () {
        return this.mapping.clone();
    };
    NumbersColumn.prototype.setMapping = function (mapping) {
        if (this.mapping.eq(mapping)) {
            return;
        }
        this.fire([NumbersColumn_1.EVENT_MAPPING_CHANGED, Column.EVENT_DIRTY_HEADER, Column.EVENT_DIRTY_VALUES, Column.EVENT_DIRTY], this.mapping.clone(), this.mapping = mapping);
    };
    NumbersColumn.prototype.getColor = function (row) {
        return NumberColumn.prototype.getColor.call(this, row);
    };
    NumbersColumn.prototype.getColorMapping = function () {
        return this.colorMapping.clone();
    };
    NumbersColumn.prototype.setColorMapping = function (mapping) {
        if (this.colorMapping.eq(mapping)) {
            return;
        }
        this.fire([NumbersColumn_1.EVENT_COLOR_MAPPING_CHANGED, Column.EVENT_DIRTY_HEADER, Column.EVENT_DIRTY_VALUES, Column.EVENT_DIRTY], this.colorMapping.clone(), this.colorMapping = mapping);
    };
    NumbersColumn.prototype.isFiltered = function () {
        return NumberColumn.prototype.isFiltered.call(this);
    };
    NumbersColumn.prototype.getFilter = function () {
        return NumberColumn.prototype.getFilter.call(this);
    };
    NumbersColumn.prototype.setFilter = function (value) {
        if (value === void 0) { value = { min: -Infinity, max: +Infinity, filterMissing: false }; }
        NumberColumn.prototype.setFilter.call(this, value);
    };
    NumbersColumn.prototype.filter = function (row) {
        return NumberColumn.prototype.filter.call(this, row);
    };
    var NumbersColumn_1;
    NumbersColumn.EVENT_MAPPING_CHANGED = NumberColumn.EVENT_MAPPING_CHANGED;
    NumbersColumn.EVENT_COLOR_MAPPING_CHANGED = NumberColumn.EVENT_COLOR_MAPPING_CHANGED;
    NumbersColumn.EVENT_SORTMETHOD_CHANGED = NumberColumn.EVENT_SORTMETHOD_CHANGED;
    NumbersColumn.EVENT_FILTER_CHANGED = NumberColumn.EVENT_FILTER_CHANGED;
    NumbersColumn.CENTER = 0;
    NumbersColumn = NumbersColumn_1 = __decorate([
        toolbar('filterNumber', 'colorMapped', 'editMapping'),
        dialogAddons('sort', 'sortNumbers'),
        SortByDefault('descending')
    ], NumbersColumn);
    return NumbersColumn;
}(ArrayColumn));
export default NumbersColumn;
