// lineupjs/src/ui/dialogs/BooleanFilterDialog.js
import { __extends } from "tslib";
import ADialog from './ADialog';
import { updateFilterState, uniqueId } from './utils';
var BooleanFilterDialog = (function (_super) {
    __extends(BooleanFilterDialog, _super);
    function BooleanFilterDialog(column, dialog) {
        var _this = _super.call(this, dialog, {
            fullDialog: true
        }) || this;
        _this.column = column;
        _this.before = _this.column.getFilter();
        return _this;
    }
    BooleanFilterDialog.prototype.build = function (node) {
        var id = uniqueId(this.dialog.idPrefix);
        node.insertAdjacentHTML('beforeend', "\n     <div class=\"lu-checkbox\"><input id=\"" + id + "0\" type=\"radio\" name=\"boolean_check\" value=\"null\" " + (this.before == null ? 'checked="checked"' : '') + "><label id=\"" + id + "0\">No Filter</label></div>\n     <div class=\"lu-checkbox\"><input id=\"" + id + "1\" type=\"radio\" name=\"boolean_check\" value=\"true\" " + (this.before === true ? 'checked="checked"' : '') + "><label id=\"" + id + "1\">True</label></div>\n     <div class=\"lu-checkbox\"><input id=\"" + id + "2\" type=\"radio\" name=\"boolean_check\" value=\"false\" " + (this.before === false ? 'checked="checked"' : '') + "><label id=\"" + id + "2\">False</label></div>\n    ");
    };
    BooleanFilterDialog.prototype.updateFilter = function (filter) {
        updateFilterState(this.attachment, this.column, filter != null);
        this.column.setFilter(filter);
    };
    BooleanFilterDialog.prototype.reset = function () {
        var v = 'null';
        this.forEach('input[type="radio"]', function (d) { return d.checked = d.value === v; });
        this.updateFilter(null);
    };
    BooleanFilterDialog.prototype.submit = function () {
        var isTrue = this.findInput('input[type="radio"][value="true"]').checked;
        var isFalse = this.findInput('input[type="radio"][value="false"]').checked;
        this.updateFilter(isTrue ? true : (isFalse ? false : null));
        return true;
    };
    return BooleanFilterDialog;
}(ADialog));
export default BooleanFilterDialog;
