// lineupjs/src/ui/taggle/spaceFillingRule.js
import { isGroup } from '../../model';
export default function spaceFillingRule(config) {
    function levelOfDetail(item, height) {
        var group = isGroup(item);
        var maxHeight = group ? config.groupHeight : config.rowHeight;
        if (height >= maxHeight * 0.9) {
            return 'high';
        }
        return 'low';
    }
    function itemHeight(data, availableHeight, selection) {
        var visibleHeight = availableHeight - config.rowHeight - 5;
        var items = data.filter(function (d) { return !isGroup(d); });
        var groups = data.length - items.length;
        var lastItems = items.reduce(function (a, d) { return a + (d.meta === 'last' || d.meta === 'first last' ? 1 : 0); }, 0);
        var selected = items.reduce(function (a, d) { return a + (selection.has(d.i) ? 1 : 0); }, 0);
        var unselected = items.length - selected;
        var groupSeparators = groups + lastItems;
        if (unselected <= 0) {
            return { height: config.rowHeight, violation: '' };
        }
        var available = visibleHeight - groups * config.groupHeight - groupSeparators * config.groupPadding - selected * config.rowHeight;
        var height = available / unselected;
        if (height < 1) {
            return {
                height: 1,
                violation: "Not possible to fit all rows on the screen. Set filters or aggregate groups to make it fit again."
            };
        }
        if (height > config.rowHeight) {
            return {
                height: config.rowHeight,
                violation: ""
            };
        }
        return { height: height, violation: '' };
    }
    return {
        apply: function (data, availableHeight, selection) {
            var _a = itemHeight(data, availableHeight, selection), violation = _a.violation, height = _a.height;
            var item = function (item) {
                if (selection.has(item.i)) {
                    return config.rowHeight;
                }
                return height;
            };
            return { item: item, group: config.groupHeight, violation: violation };
        },
        levelOfDetail: levelOfDetail
    };
}
