// lineupjs/src/ui/panel/SidePanelRanking.js
import { suffix } from '../../internal/AEventDispatcher';
import { isSupportType } from '../../model';
import Ranking from '../../model/Ranking';
import Hierarchy from './Hierarchy';
import SidePanelEntryVis from './SidePanelEntryVis';
import { dialogContext } from '../toolbar';
import MoreRankingOptionsDialog from '../dialogs/MoreRankingOptionsDialog';
var SidePanelRanking = (function () {
    function SidePanelRanking(ranking, ctx, document, options) {
        this.ranking = ranking;
        this.ctx = ctx;
        this.options = options;
        this.entries = new Map();
        this.node = document.createElement('section');
        this.header = document.createElement('div');
        this.dropdown = document.createElement('div');
        this.dropdown.innerHTML = this.header.innerHTML = "<span>" + ranking.getLabel() + "</span><i class=\"lu-action\" title=\"More &hellip;\"><span aria-hidden=\"true\">More &hellip;\"></span></i>";
        this.header.lastElementChild.onclick = this.dropdown.lastElementChild.onclick = function (evt) {
            evt.stopPropagation();
            var dialog = new MoreRankingOptionsDialog(ranking, dialogContext(ctx, 1, evt), ctx);
            dialog.open();
        };
        this.hierarchy = this.options.hierarchy ? new Hierarchy(ctx, document) : null;
        this.init();
    }
    SidePanelRanking.prototype.init = function () {
        var _this = this;
        this.node.innerHTML = "\n      <div><main></main></div>\n    ";
        if (this.hierarchy) {
            this.node.insertBefore(this.hierarchy.node, this.node.firstChild);
        }
        if (this.hierarchy) {
            this.ranking.on(suffix('.panel', Ranking.EVENT_GROUP_CRITERIA_CHANGED, Ranking.EVENT_SORT_CRITERIA_CHANGED, Ranking.EVENT_GROUP_SORT_CRITERIA_CHANGED), function () {
                _this.updateHierarchy();
            });
        }
        this.ranking.on(suffix('.panel', Ranking.EVENT_ADD_COLUMN, Ranking.EVENT_REMOVE_COLUMN, Ranking.EVENT_MOVE_COLUMN), function () {
            _this.updateList();
            _this.updateHierarchy();
        });
        this.ranking.on(suffix('.panel', Ranking.EVENT_LABEL_CHANGED), function () {
            _this.dropdown.firstElementChild.textContent = _this.header.firstElementChild.textContent = _this.ranking.getLabel();
        });
    };
    Object.defineProperty(SidePanelRanking.prototype, "active", {
        get: function () {
            return this.node.classList.contains('lu-active');
        },
        set: function (value) {
            this.node.classList.toggle('lu-active', value);
            this.header.classList.toggle('lu-active', value);
            this.dropdown.classList.toggle('lu-active', value);
            if (value) {
                return;
            }
            this.updateList();
            this.updateHierarchy();
        },
        enumerable: true,
        configurable: true
    });
    SidePanelRanking.prototype.update = function (ctx) {
        this.ctx = ctx;
        this.updateList();
        this.updateHierarchy();
    };
    SidePanelRanking.prototype.updateHierarchy = function () {
        if (!this.hierarchy || !this.active) {
            return;
        }
        this.hierarchy.update(this.ranking);
    };
    SidePanelRanking.prototype.updateList = function () {
        var _this = this;
        if (!this.active) {
            return;
        }
        var node = this.node.querySelector('main');
        var columns = this.ranking.flatColumns.filter(function (d) { return !isSupportType(d); });
        if (columns.length === 0) {
            node.innerHTML = '';
            this.entries.forEach(function (d) { return d.destroy(); });
            this.entries.clear();
            return;
        }
        node.innerHTML = "";
        var copy = new Map(this.entries);
        this.entries.clear();
        columns.forEach(function (col) {
            var existing = copy.get(col.id);
            if (existing) {
                existing.update(_this.ctx);
                node.appendChild(existing.node);
                _this.entries.set(col.id, existing);
                copy.delete(col.id);
                return;
            }
            console.assert(node.ownerDocument != null);
            var entry = new SidePanelEntryVis(col, _this.ctx, node.ownerDocument);
            node.appendChild(entry.node);
            _this.entries.set(col.id, entry);
        });
        copy.forEach(function (d) { return d.destroy(); });
    };
    SidePanelRanking.prototype.destroy = function () {
        this.header.remove();
        this.node.remove();
        this.ranking.on(suffix('.panel', Ranking.EVENT_GROUP_CRITERIA_CHANGED, Ranking.EVENT_SORT_CRITERIA_CHANGED, Ranking.EVENT_GROUP_SORT_CRITERIA_CHANGED, Ranking.EVENT_ADD_COLUMN, Ranking.EVENT_MOVE_COLUMN, Ranking.EVENT_REMOVE_COLUMN, Ranking.EVENT_LABEL_CHANGED), null);
        this.entries.forEach(function (d) { return d.destroy(); });
        this.entries.clear();
    };
    return SidePanelRanking;
}());
export default SidePanelRanking;
