// lineupjs/src/internal/AEventDispatcher.js
import { __spreadArrays } from "tslib";
import { dispatch } from 'd3-dispatch';
export function suffix(suffix) {
    var prefix = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        prefix[_i - 1] = arguments[_i];
    }
    return prefix.map(function (p) { return "" + p + suffix; });
}
var AEventDispatcher = (function () {
    function AEventDispatcher() {
        var events = this.createEventList();
        this.listenerEvents = new Set(events);
        this.listeners = dispatch.apply(void 0, events);
        var that = this;
        this.forwarder = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            that.fireImpl.apply(that, __spreadArrays([this.type, this.primaryType, this.origin], args));
        };
    }
    AEventDispatcher.prototype.on = function (type, listener) {
        var _this = this;
        if (Array.isArray(type)) {
            type.forEach(function (d) {
                if (_this.listenerEvents.has(d.split('.')[0])) {
                    _this.listenersChanged(d, Boolean(listener));
                    _this.listeners.on(d, listener);
                }
                else if (__DEBUG__ && !d.includes('.')) {
                    console.warn(_this, 'invalid event type', d);
                }
            });
        }
        else if (this.listenerEvents.has(type.split('.')[0])) {
            this.listenersChanged(type, Boolean(listener));
            this.listeners.on(type, listener);
        }
        else if (__DEBUG__ && !type.includes('.')) {
            console.warn(this, 'invalid event type', type);
        }
        return this;
    };
    AEventDispatcher.prototype.listenersChanged = function (_type, _active) {
    };
    AEventDispatcher.prototype.createEventList = function () {
        return [];
    };
    AEventDispatcher.prototype.fire = function (type) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var primaryType = Array.isArray(type) ? type[0] : type;
        this.fireImpl.apply(this, __spreadArrays([type, primaryType, this], args));
    };
    AEventDispatcher.prototype.fireImpl = function (type, primaryType, origin) {
        var _this = this;
        var args = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            args[_i - 3] = arguments[_i];
        }
        var fireImpl = function (t) {
            if (!_this.listenerEvents.has(t)) {
                if (__DEBUG__) {
                    console.warn(_this, 'invalid event type', t);
                }
                return;
            }
            var context = {
                source: _this,
                origin: origin,
                type: t,
                primaryType: primaryType,
                args: args
            };
            _this.listeners.apply(t, context, args);
        };
        if (Array.isArray(type)) {
            type.forEach(fireImpl.bind(this));
        }
        else {
            fireImpl(type);
        }
    };
    AEventDispatcher.prototype.forward = function (from) {
        var types = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            types[_i - 1] = arguments[_i];
        }
        from.on(types, this.forwarder);
    };
    AEventDispatcher.prototype.unforward = function (from) {
        var types = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            types[_i - 1] = arguments[_i];
        }
        from.on(types, null);
    };
    return AEventDispatcher;
}());
export default AEventDispatcher;
