// lineupjs/src/renderer/CircleCellRenderer.js
import { isNumberColumn } from '../model';
import { isNumbersColumn } from '../model/INumberColumn';
import { colorOf } from './impose';
import { ERenderMode } from './interfaces';
import { renderMissingDOM } from './missing';
import { attr, noop, noRenderer, setText } from './utils';
var CircleCellRenderer = (function () {
    function CircleCellRenderer() {
        this.title = 'Proportional Symbol';
    }
    CircleCellRenderer.prototype.canRender = function (col, mode) {
        return isNumberColumn(col) && mode === ERenderMode.CELL && !isNumbersColumn(col);
    };
    CircleCellRenderer.prototype.create = function (col, _context, _hist, imposer) {
        return {
            template: "<div style=\"background: radial-gradient(circle closest-side, red 100%, transparent 100%)\" title=\"\">\n              <div class=\"lu-hover-only\"></div>\n          </div>",
            update: function (n, d) {
                var v = col.getNumber(d);
                var p = Math.round(v * 100);
                var missing = renderMissingDOM(n, col, d);
                attr(n, {}, {
                    background: missing ? null : "radial-gradient(circle closest-side, " + colorOf(col, d, imposer) + " " + p + "%, transparent " + p + "%)"
                });
                setText(n.firstElementChild, col.getLabel(d));
            },
            render: noop
        };
    };
    CircleCellRenderer.prototype.createGroup = function () {
        return noRenderer;
    };
    CircleCellRenderer.prototype.createSummary = function () {
        return noRenderer;
    };
    return CircleCellRenderer;
}());
export default CircleCellRenderer;
