// lineupjs/src/renderer/ImageCellRenderer.js
import { ERenderMode } from './interfaces';
import { renderMissingDOM } from './missing';
import { noop, noRenderer } from './utils';
import LinkColumn from '../model/LinkColumn';
var ImageCellRenderer = (function () {
    function ImageCellRenderer() {
        this.title = 'Image';
    }
    ImageCellRenderer.prototype.canRender = function (col, mode) {
        return col instanceof LinkColumn && mode === ERenderMode.CELL;
    };
    ImageCellRenderer.prototype.create = function (col) {
        return {
            template: "<div></div>",
            update: function (n, d) {
                var missing = renderMissingDOM(n, col, d);
                var v = col.getLink(d);
                n.title = v ? v.alt : '';
                n.style.backgroundImage = missing || !v ? null : "url('" + v.href + "')";
            },
            render: noop
        };
    };
    ImageCellRenderer.prototype.createGroup = function () {
        return noRenderer;
    };
    ImageCellRenderer.prototype.createSummary = function () {
        return noRenderer;
    };
    return ImageCellRenderer;
}());
export default ImageCellRenderer;
