// lineupjs/src/renderer/RankCellRenderer.js
import RankColumn from '../model/RankColumn';
import { renderMissingDOM } from './missing';
import { noop, noRenderer, setText } from './utils';
var RankCellRenderer = (function () {
    function RankCellRenderer() {
        this.title = 'Default';
    }
    RankCellRenderer.prototype.canRender = function (col) {
        return col instanceof RankColumn;
    };
    RankCellRenderer.prototype.create = function (col) {
        return {
            template: "<div class=\"lu-right\"> </div>",
            update: function (n, d) {
                renderMissingDOM(n, col, d);
                setText(n, col.getLabel(d));
            },
            render: noop
        };
    };
    RankCellRenderer.prototype.createGroup = function (col) {
        return {
            template: "<div><div></div><div></div></div>",
            update: function (n, _group, rows) {
                var fromTSpan = n.firstElementChild;
                var toTSpan = n.lastElementChild;
                if (rows.length === 0) {
                    fromTSpan.textContent = '';
                    toTSpan.textContent = '';
                    return;
                }
                fromTSpan.textContent = col.getLabel(rows[0]);
                toTSpan.textContent = col.getLabel(rows[rows.length - 1]);
            }
        };
    };
    RankCellRenderer.prototype.createSummary = function () {
        return noRenderer;
    };
    return RankCellRenderer;
}());
export default RankCellRenderer;
