// lineupjs/src/builder/column/ActionsColumnBuilder.js
import { __extends } from "tslib";
import ColumnBuilder from './ColumnBuilder';
var ActionsColumnBuilder = (function (_super) {
    __extends(ActionsColumnBuilder, _super);
    function ActionsColumnBuilder() {
        return _super.call(this, 'actions', '') || this;
    }
    ActionsColumnBuilder.prototype.action = function (action) {
        return this.actions([action]);
    };
    ActionsColumnBuilder.prototype.actions = function (actions) {
        var _a;
        if (!this.desc.actions) {
            this.desc.actions = [];
        }
        (_a = this.desc.actions).push.apply(_a, actions);
        return this;
    };
    ActionsColumnBuilder.prototype.groupAction = function (action) {
        return this.groupActions([action]);
    };
    ActionsColumnBuilder.prototype.groupActions = function (actions) {
        var _a;
        if (!this.desc.groupActions) {
            this.desc.groupActions = [];
        }
        (_a = this.desc.groupActions).push.apply(_a, actions);
        return this;
    };
    return ActionsColumnBuilder;
}(ColumnBuilder));
export default ActionsColumnBuilder;
export function buildActionsColumn() {
    return new ActionsColumnBuilder();
}
