// lineupjs/src/builder/column/ColumnBuilder.js
var ColumnBuilder = (function () {
    function ColumnBuilder(type, column) {
        this.desc = { column: column, type: type, label: column ? column[0].toUpperCase() + column.slice(1) : type };
    }
    ColumnBuilder.prototype.label = function (label) {
        this.desc.label = label;
        return this;
    };
    ColumnBuilder.prototype.description = function (description) {
        this.desc.description = description;
        return this;
    };
    ColumnBuilder.prototype.frozen = function () {
        this.desc.frozen = true;
        return this;
    };
    ColumnBuilder.prototype.renderer = function (renderer, groupRenderer, summaryRenderer) {
        if (renderer) {
            this.desc.renderer = renderer;
        }
        if (groupRenderer) {
            this.desc.groupRenderer = groupRenderer;
        }
        if (summaryRenderer) {
            this.desc.summaryRenderer = summaryRenderer;
        }
        return this;
    };
    ColumnBuilder.prototype.custom = function (key, value) {
        this.desc[key] = value;
        return this;
    };
    ColumnBuilder.prototype.width = function (width) {
        this.desc.width = width;
        return this;
    };
    ColumnBuilder.prototype.color = function (_color) {
        return this;
    };
    ColumnBuilder.prototype.asArray = function (labels) {
        console.assert(['boolean', 'categorical', 'date', 'number', 'string', 'link'].includes(this.desc.type));
        this.desc.type += 's';
        var a = this.desc;
        if (Array.isArray(labels)) {
            a.labels = labels;
            a.dataLength = labels.length;
        }
        else if (typeof labels === 'number') {
            a.dataLength = labels;
        }
        return this;
    };
    ColumnBuilder.prototype.asMap = function () {
        console.assert(['categorical', 'date', 'number', 'string', 'link'].includes(this.desc.type));
        this.desc.type += 'Map';
        return this;
    };
    ColumnBuilder.prototype.build = function (_data) {
        return this.desc;
    };
    return ColumnBuilder;
}());
export default ColumnBuilder;
export function buildColumn(type, column) {
    return new ColumnBuilder(type, column);
}
