// lineupjs/src/builder/column/StringColumnBuilder.js
import { __extends } from "tslib";
import ColumnBuilder from './ColumnBuilder';
var StringColumnBuilder = (function (_super) {
    __extends(StringColumnBuilder, _super);
    function StringColumnBuilder(column) {
        return _super.call(this, 'string', column) || this;
    }
    StringColumnBuilder.prototype.editable = function () {
        this.desc.type = 'annotate';
        return this;
    };
    StringColumnBuilder.prototype.alignment = function (align) {
        this.desc.alignment = align;
        return this;
    };
    StringColumnBuilder.prototype.html = function () {
        this.desc.escape = false;
        return this;
    };
    StringColumnBuilder.prototype.pattern = function (pattern, templates) {
        this.desc.type = 'link';
        this.desc.pattern = pattern;
        if (templates) {
            this.desc.patternTemplates = templates;
        }
        return this;
    };
    return StringColumnBuilder;
}(ColumnBuilder));
export default StringColumnBuilder;
export function buildStringColumn(column) {
    return new StringColumnBuilder(column);
}
